contracts/libraries/LibArray.sol

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+/**
+ * @title LibArray
+ * @dev A library for array-related utility functions in Solidity.
+ */
+library LibArray {
+  /**
+   * @dev Calculates the sum of an array of uint256 values.
+   * @param data The array of uint256 values for which the sum is calculated.
+   * @return result The sum of the provided array.
+   */
+  function sum(uint256[] memory data) internal pure returns (uint256 result) {
+    assembly ("memory-safe") {
+      // Load the length (first 32 bytes)
+      let len := mload(data)
+      let dataElementLocation := add(data, 0x20)
+
+      // Iterate until the bound is not met.
+      for { let end := add(dataElementLocation, mul(len, 0x20)) } lt(dataElementLocation, end) {
+        dataElementLocation := add(dataElementLocation, 0x20)
+      } { result := add(result, mload(dataElementLocation)) }
+    }
+  }
+
+  function findNormalizedSumAndUpperBound(
```

Comment:
add docstring

Reply:
state in the docstring:
- this modifies input array
- idea of the sum: all elements in the normalized array are less than `1/divisor`

Reply:
consider that `inplace-` to the method name

Reply:
this function does not modify input array